stack/stack_test.go

```diff
@@ -0,0 +1,92 @@
+// Copyright (c) Alex Ellis 2017. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+package stack
+
+import (
+	"testing"
+	"reflect"
+	"sort"
+)
+
+var ParseYamlTests= []struct {
+	title        string
+	filter       string
+	functions    []string
+}{
+	{
+		title:        "Filter functions only containing 'node'",
+		filter:       "node",
+		functions:    []string{"nodejs-echo"},
+	},
+	{
+		title:        "Filter functions only containing 'echo'",
+		filter:       "echo",
+		functions:    []string{"nodejs-echo", "ruby-echo"},
+	},
+	{
+		title:        "Filter functions only containing '.+-.+'",
+		filter:       ".+-.+",
+		functions:    []string{"nodejs-echo", "ruby-echo", "url-ping"},
+	},
+	{
+		title:        "Filter for all functions: '.*'",
+		filter:       ".*",
+		functions:    []string{"imagemagick", "nodejs-echo", "ruby-echo", "url-ping"},
+	},
```

Comment:
You should add some negative tests, for example what happens when the filter results in an empty function slice.

Reply:
Also what happens when the regex is invalid.

Reply:
When the function results in no functions being matched, the same behavior will be exhibited as if you had called `faas-cli <verb> -f <YAML>` on a YAML file which does not have a `functions:` object. What it does with the `build` action is prints the following error message.

`Please provide a valid -image name for your Docker image.`

I don't think it's within the scope of this PR to fix the behavior of commands when they parse a YAML file and there are no functions.

If the regex is invalid, the err variable will be set (I will change it to storing and checking the `err` variable after the call to `regexp.MatchString()`)

Reply:
Actually, I was wrong about the error behavior. I found a case that causes the `err` var to be set to non-`nil`:

```
faas-cli deploy -f samples.yml --filter "*"
2017/09/06 06:32:49 Error parsing regex passed in through --filter argument: error parsing regexp: missing argument to repetition operator: `*`
2017/09/06 06:32:49 error parsing regexp: missing argument to repetition operator: `*`
```

If you issue a command with quotes around the `*` variable without an argument, the go `regexp` parser errors out. I think this is another reason why filter should be wildcard matching rather than regexp matching. I think we can use [this class for wildcard matching](https://github.com/ryanuber/go-glob/blob/master/glob.go).

Reply:
@ems5311  we've had a chat about this, and think your glob suggestion is a good one, we'd still like to keep the regex handling however so how about two flags:
- `--filter=` using your glob suggestion for a simple filtering experience
- `--regex=` using the current regex handling that you have here for a richer filtering

Thanks for all your work on this btw, its looking great so far !!!